Cross-Document View Transitions cause permanent rendering suppression when loading from BFCache.

mattwoodrow opened the pull request:
#### d9d731244a3745f91f3a49e0875a4350da050d8a
<pre>
Cross-Document View Transitions cause permanent rendering suppression when loading from BFCache.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280703">https://bugs.webkit.org/show_bug.cgi?id=280703</a>
&lt;<a href="https://rdar.apple.com/137081148">rdar://137081148</a>&gt;

Reviewed by Tim Nguyen.

Three changes to fix this (and the associated WPT):

Clear m_hasBeenRevealed on Document when reactivating it from BFCache, so that
the pagereveal event gets fired again.

Skip any active view transitions when suspending a Document to put it into the
BFCache.

Start rendering suppression for a cross-document view transition immediately,
instead of waiting until the next rendering update. This is because we
(incorrectly) also immediately capture the old state.

Duplicates the old_vt_promises_bfcache.html WPT into the local wpt folder, so
that the &apos;UsesBackForwardCache&apos; attribute can be applied to make the test run
correctly.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/css/css-view-transitions/navigation/old_vt_promises_bfcache-expected.txt: Added.
* LayoutTests/http/wpt/css/css-view-transitions/navigation/old_vt_promises_bfcache.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/old_vt_promises_bfcache-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearRevealForReactivation):
(WebCore::Document::suspend):
(WebCore::Document::setRenderingIsSuppressedForViewTransitionImmediately):
(WebCore::Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::setupViewTransition):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::firePageShowEvent):

Canonical link: <a href="https://commits.webkit.org/285582@main">https://commits.webkit.org/285582@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73f2ebad81224b8784156383b9b5f2965b1abcdc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77331 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24351 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75228 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/324 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57472 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15948 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47478 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62922 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37884 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44117 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20393 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22680 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20748 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78992 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65907 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65187 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16113 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7172 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/392 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3129 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/421 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/406 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -7577,6 +7577,9 @@ imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-m
 imported/w3c/web-platform-tests/css/css-view-transitions/navigation/chromium-paint-holding-timeout.html [ ImageOnlyFailure ]
 webkit.org/b/278756 imported/w3c/web-platform-tests/css/css-view-transitions/navigation/skip-outbound-vt-before-reveal.html [ Failure ]
 
+# BFCache not enabled in WebKitTestRunner
```
maybe add a comment saying this is duplicated in `http/wpt`

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10894,6 +10903,13 @@ void Document::setRenderingIsSuppressedForViewTransitionAfterUpdateRendering()
     m_enableRenderingIsSuppressedForViewTransitionAfterUpdateRendering = true;
 }
 
+void Document::setRenderingIsSuppressedForViewTransitionImmediately()
```
Should we add a FIXME to remove this once we fix the timing of capturing the old state?

mattwoodrow committed:
Cross-Document View Transitions cause permanent rendering suppression when loading from BFCache.
https://bugs.webkit.org/show_bug.cgi?id=280703
<rdar://137081148>

Reviewed by Tim Nguyen.

Three changes to fix this (and the associated WPT):

Clear m_hasBeenRevealed on Document when reactivating it from BFCache, so that
the pagereveal event gets fired again.

Skip any active view transitions when suspending a Document to put it into the
BFCache.

Start rendering suppression for a cross-document view transition immediately,
instead of waiting until the next rendering update. This is because we
(incorrectly) also immediately capture the old state.

Duplicates the old_vt_promises_bfcache.html WPT into the local wpt folder, so
that the 'UsesBackForwardCache' attribute can be applied to make the test run
correctly.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/css/css-view-transitions/navigation/old_vt_promises_bfcache-expected.txt: Added.
* LayoutTests/http/wpt/css/css-view-transitions/navigation/old_vt_promises_bfcache.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/old_vt_promises_bfcache-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearRevealForReactivation):
(WebCore::Document::suspend):
(WebCore::Document::setRenderingIsSuppressedForViewTransitionImmediately):
(WebCore::Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::setupViewTransition):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::firePageShowEvent):

Canonical link: https://commits.webkit.org/285582@main
